### PR TITLE
fix(python): PR-2 monitor contract alignment and UTC datetimes

### PR DESCRIPTION
**Pull request by tomazb**
## Summary

- Align `Monitor` with typed `K8sConfig`/`TrueNASConfig` factories and real `get_*` client APIs (replacing stale `list_*` calls and dict-shaped assumptions).
- Normalize `kubernetes` → `openshift` config sections at load time; add shared UTC datetime helpers and fix naive/aware comparison bugs in the K8s client.
- Update unit tests to catch contract drift (dataclass fixtures, timezone edge cases, factory wiring).

**Spec:** [docs/superpowers/specs/2026-05-28-pr-2-python-correctness-design.md](docs/superpowers/specs/2026-05-28-pr-2-python-correctness-design.md)

**Plan item:** PR 2 in [docs/superpowers/plans/2026-05-28-repo-health-remediation.md](docs/superpowers/plans/2026-05-28-repo-health-remediation.md)

## Mandatory PR checklist

- [x] **Scope:** Maps to one plan item (PR 2).
- [x] **Correctness:** Added/updated tests for changed logic.
- [x] **Regression Safety:** Relevant existing tests pass (76/76).
- [x] **Security:** No insecure defaults introduced; secrets not exposed.
- [x] **Reliability:** Failure paths handled; datetime comparisons are UTC-aware.
- [x] **Performance:** No obvious unbounded loops/maps or avoidable N^2 regressions.
- [ ] **Docs:** Config behavior unchanged in user-facing YAML; deploy manifest rename deferred to PR-8.
- [ ] **Ops/Runbook:** No operational behavior change requiring runbook update.
- [x] **Verification Evidence:** See test plan below.
- [x] **Tracking:** Plan status/checklists updated.

## Test plan

- [x] `cd python && pytest tests/unit/test_config.py tests/unit/test_monitor.py tests/unit/test_k8s_client.py tests/unit/test_time_utils.py -v` — 57 passed
- [x] `make python-test` — 76 passed (coverage gate 90% still pre-existing: 76.48% vs 74.37% on main)
- [x] `mypy` on changed source files — success
- [ ] `make python-lint` — flake8 E501 pre-existing repo-wide (main also fails)
- [ ] `bandit` — not installed locally; CI to verify

## Review notes

- CodeRabbit review initiated but did not complete within session timeout; manual review performed on diff prior to commit.

## Summary by CodeRabbit

* **New Features**
  * OpenShift config with Kubernetes aliasing, API connection check, namespace listing, TrueNAS HTTPS option, and UTC-based timestamps/reporting

* **Bug Fixes**
  * Timezone-aware datetime handling, normalized comparisons, improved orphan detection and age calculations, and stricter TrueNAS URL/timeout parsing

* **Tests**
  * Expanded unit tests for config parsing, time utilities, client connectivity, orphan detection, health checks, and recommendations

* **Documentation**
  * Added design spec for PR 2 and updated remediation plan statuses

* **Refactor**
  * Monitoring workflow switched to typed clients and centralized UTC helpers

[Review Change Stack](https://app.coderabbit.ai/change-stack/tomazb/kubernetes-truenas-democratic-tool/pull/52?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: python/truenas_storage_monitor/config.py [hunk 1]
```diff
@@ -78,6 +86,72 @@ def logging(self) -> Dict[str, Any]:
...
+    normalized = url if "://" in url else f"https://{url}"
+    parsed = urlparse(normalized)
+    host = parsed.hostname or url
+    if parsed.port is not None:
+        port = parsed.port
+    elif parsed.scheme == "http":
+        port = 80
+    else:
+        port = 443
+    return host, port
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/config.py]: Fixed in a72865b: `parse_truenas_url` now wraps `urlparse` failures as `ConfigurationError` and validates hostname presence.

File: python/truenas_storage_monitor/config.py [hunk 2]
```diff
@@ -78,6 +86,72 @@ def logging(self) -> Dict[str, Any]:
...
+def parse_timeout_seconds(value: Any) -> int:
+    """Parse timeout values such as 30 or '30s' into seconds."""
+    if isinstance(value, int):
+        return value
+    if isinstance(value, str):
+        stripped = value.strip()
+        if stripped.endswith("s"):
+            return int(stripped[:-1])
+        return int(stripped)
+    raise ConfigurationError(f"Invalid timeout value: {value!r}")
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/config.py]: Fixed in a72865b: `parse_timeout_seconds` rejects bools, wraps parse failures as `ConfigurationError`, and requires > 0 seconds.

[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/config.py]: Fixed in a72865b (same change): bool rejection, wrapped parse failures, and positive-second validation in `parse_timeout_seconds`.

File: python/truenas_storage_monitor/config.py [hunk 3]
```diff
@@ -78,6 +86,72 @@ def logging(self) -> Dict[str, Any]:
...
+    """Normalize legacy ``kubernetes`` section to canonical ``openshift``."""
+    if "kubernetes" not in config:
+        return config
+
+    kubernetes = config.pop("kubernetes")
+    if "openshift" not in config:
+        config["openshift"] = kubernetes
+    else:
+        config["openshift"] = merge_configs(kubernetes, config["openshift"])
+    return config
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/config.py:176]: Fixed in a72865b: `normalize_cluster_config` validates `kubernetes`/`openshift` are mappings before merge.

File: python/truenas_storage_monitor/k8s_client.py [hunk 4]
```diff
@@ -442,7 +465,7 @@ def find_orphaned_pvs(self) -> List[OrphanedResource]:
                     name=pv.name,
                     namespace=None,
                     volume_handle=pv.volume_handle,
-                    creation_time=pv.creation_time or datetime.now(),
+                    creation_time=pv.creation_time or utc_now(),
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/k8s_client.py]: Fixed in a72865b: orphaned PV `creation_time` uses `ensure_utc(pv.creation_time)` when present.

[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/k8s_client.py]: Fixed in a72865b: same `ensure_utc` normalization for orphaned PV timestamps.

File: python/truenas_storage_monitor/monitor.py [hunk 5]
```diff
@@ -59,133 +63,126 @@ def find_orphaned_resources(
...
+
+            if not self._has_corresponding_truenas_volume(pv, truenas_volumes):
+                age = resource_age(created)
+                orphaned.append(
+                    {
+                        "name": pv.name,
+                        "age": age,
+                        "reason": "No corresponding TrueNAS volume found",
+                        "size": pv.capacity or "Unknown",
+                        "storage_class": "Unknown",
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/monitor.py]: Fixed in a72865b: added `storage_class` to `PersistentVolumeInfo`, populated from `pv.spec.storage_class_name`, used in monitor output.

File: python/truenas_storage_monitor/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/config.py]: Fixed in a72865b: `parse_truenas_url` returns `use_https`; `TrueNASConfig.base_url` uses scheme flag instead of inferring protocol from port.

File: python/truenas_storage_monitor/config.py [hunk 6]
```diff
@@ -78,6 +86,72 @@ def logging(self) -> Dict[str, Any]:
...
+            namespace=cluster.get("namespace"),
+            csi_driver=cluster.get("csi_driver", "org.democratic-csi.nfs"),
+            storage_class=cluster.get("storage_class"),
+            in_cluster=cluster.get("in_cluster", False),
+        )
+
+    def truenas_config(self) -> TrueNASConfig:
+        """Build typed TrueNAS client configuration."""
+        truenas = self.truenas
+        host, port = parse_truenas_url(truenas["url"])
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/config.py]: Fixed in a72865b: HTTPS on custom ports preserved via `use_https` (see `test_parse_truenas_url_https_custom_port`).

[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/config.py]: Fixed in a72865b: `truenas_config()` raises `ConfigurationError` when `truenas.url` is missing.

File: python/truenas_storage_monitor/k8s_client.py [hunk 7]
```diff
@@ -468,12 +491,12 @@ def find_orphaned_pvcs(self, pending_threshold_minutes: int = 60) -> List[Orphan
...
-        threshold = datetime.now(
-            tz=pvcs[0].creation_time.tzinfo if pvcs and pvcs[0].creation_time else None
-        ) - timedelta(minutes=pending_threshold_minutes)
+        threshold = utc_now() - timedelta(minutes=pending_threshold_minutes)
 
         for pvc in pvcs:
-            if pvc.phase == "Pending" and pvc.creation_time and pvc.creation_time < threshold:
+            if pvc.phase != "Pending" or pvc.creation_time is None:
+                continue
+            if ensure_utc(pvc.creation_time) < threshold:
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/k8s_client.py:501]: Fixed in a72865b: orphaned PVC results store `ensure_utc(pvc.creation_time)`.

File: python/truenas_storage_monitor/monitor.py [hunk 8]
```diff
@@ -59,133 +63,126 @@ def find_orphaned_resources(
...
-        volume_handle = pv.get("spec", {}).get("csi", {}).get("volumeHandle", "")
+        volume_handle = pv.volume_handle
         if not volume_handle:
             return False
 
-        # Look for matching TrueNAS volume
         for volume in truenas_volumes:
-            if volume.get("name") in volume_handle or volume_handle in volume.get("name", ""):
+            if volume.name in volume_handle or volume_handle in volume.name:
                 return True
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/monitor.py:173]: Not changed in this PR. Substring volume-handle matching is the same heuristic as pre-refactor `monitor.py` (not a regression). Tighter correlation is deferred to later orphan product-depth work to keep PR-2 scoped to contract/datetime alignment.

File: python/truenas_storage_monitor/monitor.py [hunk 9]
```diff
@@ -59,133 +63,126 @@ def find_orphaned_resources(
...
 
-        # Look for matching TrueNAS snapshot
         for truenas_snapshot in truenas_snapshots:
-            if (
-                snapshot_name in truenas_snapshot.get("name", "")
-                or truenas_snapshot.get("name", "") in snapshot_name
-            ):
+            names = {truenas_snapshot.name, truenas_snapshot.full_name}
+            if any(name and (snapshot_name in name or name in snapshot_name) for name in names):
                 return True
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/truenas_storage_monitor/monitor.py:186]: Not changed in this PR. Snapshot substring matching is inherited from the previous monitor implementation. Exact/delimiter-based matching is deferred to later orphan-correlation work.

File: python/tests/unit/test_config.py [hunk 10]
```diff
@@ -130,3 +134,127 @@ def test_load_config_file_not_found(self):
...
+        """Missing TrueNAS URL raises ConfigurationError."""
+        config = Config.__new__(Config)
+        config.data = {"truenas": {"api_key": "secret"}}
+
+        with pytest.raises(ConfigurationError, match="TrueNAS URL is required"):
+            config.truenas_config()
+
+    def test_normalize_cluster_config_rejects_non_mapping(self):
+        """Non-mapping kubernetes section raises ConfigurationError."""
+        with pytest.raises(ConfigurationError, match="kubernetes.*mapping"):
```
[the remark this answers is not in the document]
Comment (tomazb) [reply python/tests/unit/test_config.py]: Fixed in 18dbddd: switched to raw regex string `match=r"kubernetes.*mapping"`.
